securedrop/tests/test_alembic.py

```diff
@@ -24,32 +28,123 @@ def list_migrations(cfg_path, head):
+                          shell=True)  # nosec
+
+
+def get_schema(app):
+    with app.app_context():
+        return list(db.engine.execute(text('''
+            SELECT type, name, tbl_name, sql
+            FROM sqlite_master
+            ORDER BY type, name, tbl_name
+            ''')))
+
+
+@pytest.mark.parametrize('migration', ALL_MIGRATIONS)
 def test_alembic_migration_upgrade(alembic_config, config, migration):
     # run migrations in sequence from base -> head
     for mig in list_migrations(alembic_config, migration):
-        subprocess.check_call(('cd {} && alembic upgrade {}'
-                               .format(path.dirname(alembic_config),
-                                       mig)),
-                              shell=True)  # nosec
+        upgrade(alembic_config, mig)
 
 
-@pytest.mark.parametrize('migration',
-                         [x.split('_')[0] for x in ALL_MIGRATIONS])
+@pytest.mark.parametrize('migration', ALL_MIGRATIONS)
 def test_alembic_migration_downgrade(alembic_config, config, migration):
     # upgrade to the parameterized test case ("head")
-    subprocess.check_call(('cd {} && alembic upgrade {}'
-                           .format(path.dirname(alembic_config),
-                                   migration)),
-                          shell=True)  # nosec
+    upgrade(alembic_config, migration)
 
     # run migrations in sequence from "head" -> base
     migrations = list_migrations(alembic_config, migration)
     migrations.reverse()
 
     for mig in migrations:
-        subprocess.check_call(('cd {} && alembic downgrade {}'
-                               .format(path.dirname(alembic_config),
-                                       mig)),
-                              shell=True)  # nosec
+        downgrade(alembic_config, mig)
+
+
+@pytest.mark.parametrize('migration', ALL_MIGRATIONS)
+def test_schema_unchanged_after_up_then_downgrade(alembic_config,
+                                                  config,
+                                                  migration):
+    # Create the app here. Using a fixture will init the database.
+    app = create_app(config)
+
+    migrations = list_migrations(alembic_config, migration)
+
+    if len(migrations) > 1:
+        target = migrations[-2]
+        upgrade(alembic_config, target)
+    else:
+        # The first migration is the degenerate case where we don't need to
+        # get the database to some base state.
+        pass
+
+    original_schema = get_schema(app)
+
+    upgrade(alembic_config, '+1')
+    downgrade(alembic_config, '-1')
+
+    reverted_schema = get_schema(app)
+
+    # The initial migration is a degenerate case because it creates the table
+    # 'alembic_version', but rolling back the migration doesn't clear it.
+    if len(migrations) == 1:
+        reverted_schema = list(filter(lambda x: x[2] != 'alembic_version',
+                                      reverted_schema))
+
+    assert reverted_schema == original_schema
```

Comment:
Sweet. Since this also ensures the same FK relationships exist before and after the revert, this covers the cases where I wanted us to do upgrade, downgrade, upgrade.

Reply:
This actually doesn't *fully* cover that case because of non-SQL things like moving files around on the FS or purging old keys from the GPG keyring. We can add that test later.